docs/project/milestones.md

```diff
@@ -0,0 +1,305 @@
+        etc.
+    -   Where these build on top of other language or library designs, those are
+        transitively in-scope.
+-   Design for both Carbon use of C++ and C++ use of Carbon, including all major
+    C++ language features except for coroutines, is documented, cohesive, and
+    understandable by evaluators without placeholders.
+-   Evaluators can build and run tests of most C++ interoperability, including
+    both real-world C++ code that can be built with the latest release of Clang
+    and test Carbon code.
+    -   Gaps from the design need to be ones that don't undermine evaluation
+        confidence.
+-   Evaluators can effectively stress-test build speed and scaling with Carbon.
+-   Evaluators can build some key benchmarks that include C++ interoperation in
+    the critical path and get representative performance results.
+    -   This can in turn be a still smaller subset of all aspects of C++
+        interoperability based around what impacts interesting benchmarks.
+
+### Language features
+
+These are focused on the core _necessary_ features for us to reach a successful
+0.1 language that can address our goals. Some of these features are required
+directly by the above goals, others are required due to dependencies or
+interactions with the directly required features. However, we don't try to cover
+all of the features in full granularity here. There will be many minor
+components that are necessary for these to hold together but are not directly
+addressed. In general, unless something is explicitly described as partial or
+having exceptions, everything covered by that entry should be expected in the
+0.1 language.
+
+Another important point is that this doesn't commit Carbon to any _particular_
+design for any of these bullet points. Instead, it just means that the Carbon
+design must have _something_ that addresses each of these bullet points. That
+might be to add the named design to Carbon, but it equally might be a clear
+statement that Carbon will _not_ include this design but use some other language
+features to address its use cases and when rewriting C++ code using that feature
+into Carbon.
+
+#### Code organization and structuring
+
+-   Packages
+-   Libraries
+-   Implementation files
+-   Importing
+-   Namespaces
+
+#### Type system
+
+-   User-defined types
+    -   C++ interop: importing C++ types into Carbon, exporting Carbon types
+        into C++
+    -   Single inheritance
+        -   Virtual dispatch
+        -   C++ interop:
+            -   Support bi-directional inheritance: C++ <-> Carbon
+            -   Support hierarchy roots in both C++ and Carbon
+            -   Mappings across inheritance features: abstract, final, virtual.
+    -   Operator overloading
+        -   C++ interop:
+            -   Synthesize Carbon overloads for imported C++ types
+            -   Export Carbon overloads into C++
+    -   Sum types (discriminated unions)
```

Comment:
This is tricky, but I'm inclined to say "no" here.

I think we should try to *only* have the interop work at the pattern matching layer of sum-types, not the actual types. So that `std::variant`s can be matched in Carbon. When pattern matching lands firmly in C++, we might want to try to get Carbon sum types to work there, but I think that'll be post-0.1 just due to the timeline of C++ pattern matching landing.

Trying to make the types *directly* interop (similar to what I suspect we will do for integers and maybe even optionals) is I think a mistake for `std::variant` specifically -- there is deep complexity in how C++ standardized these and I'm not confident that we could get the ABI we would want in Carbon and still be *conforming* in C++. So my inclination is to try to keep these separate types that each do their own thing, and get as much use-level compatibility via pattern matching as we can w/o unifying the types themselves.

Does that make sense? Does the location of the compat in the list make sense or do you think this needs to be spelled out more somewhere?